libsolidity/lsp/LanguageServer.cpp

```
@@ -0,0 +1,354 @@
+		if (!location || !location->sourceName)
+			// LSP only has diagnostics applied to individual files.
+			continue;
+
+		Json::Value jsonDiag;
+		jsonDiag["source"] = "solc";
+		jsonDiag["severity"] = toDiagnosticSeverity(error->type());
+		jsonDiag["code"] = Json::UInt64{error->errorId().error};
+		string message = error->typeName() + ":";
+		if (string const* comment = error->comment())
+			message += " " + *comment;
+		jsonDiag["message"] = move(message);
+		jsonDiag["range"] = toRange(*location);
+
+		if (auto const* secondary = error->secondarySourceLocation())
+			for (auto&& [secondaryMessage, secondaryLocation]: secondary->infos)
+			{
+				Json::Value jsonRelated;
+				jsonRelated["message"] = secondaryMessage;
+				jsonRelated["location"] = toJson(secondaryLocation);
+				jsonDiag["relatedInformation"].append(jsonRelated);
+			}
+
+		diagnosticsBySourceUnit[*location->sourceName].append(jsonDiag);
+	}
+
+	for (string const& sourceUnitName: m_fileReader.sourceCodes() | ranges::views::keys)
+	{
+		Json::Value params;
+		params["uri"] = sourceUnitNameToClientPath(sourceUnitName);
+		params["diagnostics"] = move(diagnosticsBySourceUnit.at(sourceUnitName));
+		m_client.notify("textDocument/publishDiagnostics", move(params));
+	}
+}
+
+bool LanguageServer::run()
+{
+	while (!m_exitRequested && !m_client.closed())
+	{
+		optional<Json::Value> const jsonMessage = m_client.receive();
+		if (!jsonMessage)
+		{
+			m_client.error({}, ErrorCode::ParseError, "Error parsing JSONRPC request.");
+			continue;
+		}
+
+		try
+		{
+			string const methodName = (*jsonMessage)["method"].asString();
+			MessageID const id = (*jsonMessage)["id"];
+
+			if (auto handler = valueOrDefault(m_handlers, methodName))
+				handler(id, (*jsonMessage)["params"]);
+			else
+				m_client.error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+		}
+		catch (exception const& e)
+		{
+			m_client.error({}, ErrorCode::InternalError, "Unhandled exception: "s + e.what());
```

How is this usually presented to the user by the client? Our exception messages are often not very informative so if multi-line messages are allowed I think it would be better to return the diagnostic info from Boost like we do on the console. For example assertion messages were just empty in most cases until very recently and even now they say nothing beyond the fact that some assertion failed somewhere. This is not enough to even guess what's happening if a user reports that in an issue.

---

These are all cases where some part of the protocol is not correctly implemented. I think for a first release it is fine to keep them short. If this will result in bug reports being complicated to dig into, we can extend it later on.

---

But getting diagnostic info from boost is a one-liner and would immediately give us much more info. I don't see why we should not make it more informative from the start if it's easy. Bad error handling that swallows too much info is very annoying.